### PR TITLE
Forgot password

### PR DESCRIPTION
**Pull request by UditBhardwaj**

### CODE REVIEW

File: profiles/urls.py [hunk 1]
```diff
@@ -1,9 +1,10 @@
 from django.urls import path
 
-from profiles.views.profile import Dashboard,EditUserProfileView,UpdatePassword
+from profiles.views.profile import Dashboard,EditUserProfileView,UpdatePassword,ForgotPassword
```
Comment (chayan007) [inline profiles/urls.py:3]: leave a space after `,` and sort it lexicographically

File: profiles/views/profile.py [hunk 2]
```diff
@@ -1,13 +1,24 @@
...
+from django.contrib.auth.models import User
 from profiles.models import Profile 
-from profiles.forms import ProfileForm,EditProfileForm
+from profiles.forms import ProfileForm,EditProfileForm,ResetPassword
 from profiles.forms import ChangePasswordForm
 
 from tracker.log_to_tracker import log_to_tracker
 
+def ForgotPassword(request, username, otp):
+    if request.method == 'POST':
```
Comment (chayan007) [inline profiles/views/profile.py:13]: Please use CBVs. Add get() and post() both.

Please refrain from doing this, it's considered beginner coding.

File: templates/v2/pages/protected/reset_password.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{% extends "v2/layouts/edit-layout.html" %}
+{% block content %}
+
+<form method = "post">
```
Comment (chayan007) [inline templates/v2/pages/protected/reset_password.html:4]: attach screenshot of the template.

**Comment by chayan007** [comment]
Flow:

1. View which asks users if they have forgotten password and asks them to enter registered email.
2. Once they enters email, we check if profile exists with that email/username.
---
3. If it exists, we generate an `OTPVerification` for that profile. [layman: create a record in that model]
4. We use that profile.username and otp_object.otp for creating the URL.
5. We mail the URL to the email.
6. We click on the URL and the password reset page shows up.
7. User enters the passwords and gets into dashboard and success alert 'password has been updated'.
8. A mail goes that `Password has been updated successfully.`
---
3. Raise alert in UI that no profile exists with such email.

File: profiles/views/profile.py [hunk 4]
```diff
@@ -1,19 +1,63 @@
...
+        return render(request, 'v2/pages/protected/reset_password.html')
+
+    def post(self, request, username, otp):
+        user = User.objects.get(username=username)
+        form = ResetPassword(request.POST)
+        if form.is_valid():
+            user.set_password(form.data.get('new_password'))
+            user.save()
+            return HttpResponseRedirect('/profile/dashboard')
+        return render(request, 'v2/pages/protected/reset_password.html')
```
Comment (chayan007) [inline profiles/views/profile.py:25]: Return to the same page with saying, `Check mail for further` or something like that.

Comment (chayan007) [reply profiles/views/profile.py:25]: A simple alert box - success and danger

File: profiles/views/profile.py [hunk 5]
```diff
@@ -1,19 +1,63 @@
 from django.shortcuts import render
 from django.views import View
-from django.http import HttpResponse
+from django.http import HttpResponse,HttpResponseRedirect
 from django.contrib import messages
+from django.contrib.auth.models import User
 from profiles.models import Profile
```
Comment (chayan007) [inline profiles/views/profile.py:6]: empty line

File: profiles/views/profile.py [hunk 6]
```diff
@@ -1,19 +1,63 @@
...
 from django.views import View
-from django.http import HttpResponse
+from django.http import HttpResponse,HttpResponseRedirect
 from django.contrib import messages
+from django.contrib.auth.models import User
 from profiles.models import Profile 
-from profiles.forms import ProfileForm,EditProfileForm
+from profiles.forms import ProfileForm,EditProfileForm,ResetPassword,EnterEmail
 from profiles.forms import ChangePasswordForm
+from base import generic_mailer
```
Comment (chayan007) [inline profiles/views/profile.py:9]: empty line

File: profiles/views/profile.py [hunk 7]
```diff
@@ -1,19 +1,63 @@
...
+                messages.info(request, 'No profile exists with such email!!')
+        return render(request, 'v2/pages/protected/enter-email.html',{}) 
+
 class Dashboard(View):
 
     def get(self, request, *args, **kwargs):
         log_to_tracker(request,'dashboard')
         return render(request, 'v2/raw/dashboard.html', {})
-
+
```
Comment (chayan007) [inline profiles/views/profile.py:60]: remove space

File: profiles/views/profile.py [hunk 8]
```diff
@@ -1,19 +1,63 @@
...
+from django.http import HttpResponse,HttpResponseRedirect
 from django.contrib import messages
+from django.contrib.auth.models import User
 from profiles.models import Profile 
-from profiles.forms import ProfileForm,EditProfileForm
+from profiles.forms import ProfileForm,EditProfileForm,ResetPassword,EnterEmail
 from profiles.forms import ChangePasswordForm
+from base import generic_mailer
 
 from tracker.log_to_tracker import log_to_tracker
```
Comment (chayan007) [inline profiles/views/profile.py:12]: one more line gap

File: profiles/views/profile.py [hunk 9]
```diff
@@ -1,19 +1,63 @@
...
+            user.save()
+            return HttpResponseRedirect('/profile/dashboard')
+        return render(request, 'v2/pages/protected/reset_password.html')
+
+class EnterEmail(View):
+
+    def get(self, request, format=None):
+        return render(request, 'v2/pages/protected/enter-email.html',{}) 
+
+    def post(self, request, format=None):
```
Comment (chayan007) [inline profiles/views/profile.py:32]: what is format here for ?

File: profiles/views/profile.py [hunk 10]
```diff
@@ -1,19 +1,63 @@
...
+        form = EnterEmail(request.POST or None)
+        if form.is_valid():
+            email = form.data.get('email')
+            print(email)
+            u = User.objects.get(email = email)
+            if user is not None:
```
Comment (chayan007) [inline profiles/views/profile.py:38]: `if not user:`

**Comment by chayan007** [comment]
SOLVE ALL LINTS

File: profiles/views/profile.py (hunk 4 above)
Comment (UditBhardwaj) [reply profiles/views/profile.py:25]: this alert of "mail has been sent" is used in enter email cbv . Here user only enters new password and confirm password and then redirect to dashboard.

**Review by chayan007 — APPROVED** [review]

**Pull request merged**